### PR TITLE
TableNext with expandable functionality

### PR DESCRIPTION
**Pull request by MariaJose**
## What?
Create a new `<TableNext/>` component which has all the `Table` component's functionality + expandable functionality. This PR just focus on adding the expandable functionality that works together with Select and Drag and Drop. 

Pagination + Expandable is gonna be in another PR (Plan start working on that today)

If we want to use the expandable functionality we will need to use the `<TableNext/>` component instead, and not the `<Table/>` component. For now we are gonna use this component internally, so we don't plan to add it in the BD documentation until we add tan-stack in the component.

Tan-stack: https://tanstack.com/table/v8

## Why?
We decided to create this new component since we wanted to avoid a breaking change for now because our current Select API in our `<Table/>` component was complicating the things when adding the expandable functionality so it was necessary to change the API. 

**Pros and Cons with this approach:**

**Pros:**
- We can trying out for a bit without breaking current Table usages
- We get a similar api to tan-stack, so its prob easier to port later
- Avoid breaking change (for now)

**Cons:**
- Bigger bundle size since we now have 2 tables
- If we want to add things to the table, we now have multiple tables

**Things to keep in mind:**

- We can reduce the bundle size by not including by default and have a special import import TableNext from '@bigcommerce/big-design/table-next' or something.
- If we make a change to the table, we can temporary add the changes in both tables.
- We need to be "all-in" in terms of getting TableNext be the real Table soon.
- We have a lot of different functionality of our table component that we might need to rethink the future of the component and how we can have it be extensible.
- Prefer we consume the new API of TableNext for a bit, what if we have to do changes and end up with 2 breaking Table changes back 2 back.

The new select API it's gonna be the same as the one that tan-stack uses, since we are planning to add this library in our <TableNext/> component. 

New API for select functionality:
```js
{
0: true
0.0: true
0.1: true
1: true
1.0: true
1.1: true
2: true
2.0: true
2.1: true
}
```
Example:
`0: true` <--- this means that a parent row was selected.
`0.0: true` <--- this means that a parent's child row was selected.

API for expandable functionality:

```js
{0: true, 1: true, 2: true}
```

## Screenshots/Screen Recordings

https://user-images.githubusercontent.com/39739180/179029437-4343c25c-82e3-42a2-aa6c-40dc884b2ad6.mp4

https://user-images.githubusercontent.com/39739180/179029463-11bc738b-eb49-45b6-aedc-a5473e3a32f1.mp4

https://user-images.githubusercontent.com/39739180/179029471-4285e6f0-b695-4eb0-8427-95170ecac084.mp4

## Testing/Proof
Tests pass

### CODE REVIEW

File: packages/big-design/src/components/TableNext/Actions/Actions.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import React, { RefObject } from 'react';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/Actions/Actions.tsx:1]: Please review this file

File: packages/big-design/src/components/TableNext/DataCell/styled.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/DataCell/styled.tsx:1]: Please review this file

File: packages/big-design/src/components/TableNext/DataCell/styled.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  &:first-of-type {
+    padding-left: ${({ theme }) => theme.spacing.xLarge};
+  }
+
+  ${({ isExpandable }) =>
+    isExpandable &&
+    css`
+      padding-right: 0;
+    `}
+
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/DataCell/styled.tsx:65]: Note: There was this styled used in the `<Table/>` component, but I think that is not used 🤔 so I deleted it. Unless I'm missing something.

```css
  &:last-of-type {
    padding-right: ${({ theme }) => theme.spacing.xLarge};
  }
```

File: packages/big-design/src/components/TableNext/HeaderCell/HeaderCell.tsx [hunk 4]
```diff
@@ -0,0 +1,143 @@
+import {
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/HeaderCell/HeaderCell.tsx:1]: Please review this file

File: packages/big-design/src/components/TableNext/Row/Row.tsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
+import { ChevronRightIcon, DragIndicatorIcon, ExpandMoreIcon } from '@bigcommerce/big-design-icons';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/Row/Row.tsx:1]: Please review this file

File: packages/big-design/src/components/TableNext/Row/useRowState.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
+import { TableSelectable } from '../types';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/Row/useRowState.ts:1]: Please review this file

File: packages/big-design/src/components/TableNext/RowContainer/RowContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
+import React, { forwardRef } from 'react';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/RowContainer/RowContainer.tsx:1]: Please review this file

File: packages/big-design/src/components/TableNext/SelectAll/SelectAll.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/SelectAll/SelectAll.tsx:1]: Please review this file

File: packages/big-design/src/components/TableNext/SelectAll/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
+import { TableProps, TableSelectable } from '../types';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/SelectAll/helpers.ts:1]: Please review this file

File: packages/big-design/src/components/TableNext/SelectAll/useSelectAllState.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
+import { TableProps, TableSelectable } from '../types';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/SelectAll/useSelectAllState.ts:1]: Please review this file

File: packages/big-design/src/components/TableNext/TableNext.tsx [hunk 11]
```diff
@@ -0,0 +1,285 @@
+import React, { memo, useCallback, useRef, useState } from 'react';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/TableNext.tsx:1]: Please review this file

File: packages/big-design/src/components/TableNext/hooks/useExpandable/useExpandable.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
+import { useEffect, useState } from 'react';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/hooks/useExpandable/useExpandable.ts:1]: Please review this file

File: packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
+import { TableSelectable } from '../../types';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts:1]: Please review this file

File: packages/big-design/src/components/TableNext/hooks/useSelectable/useSelectable.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
+import { useEffect, useState } from 'react';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/hooks/useSelectable/useSelectable.ts:1]: Please review this file

File: packages/big-design/src/components/TableNext/spec.tsx [hunk 15]
```diff
@@ -0,0 +1,849 @@
+import userEvent from '@testing-library/user-event';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/spec.tsx:1]: Please review this file.

File: packages/big-design/src/components/TableNext/types.ts [hunk 16]
```diff
@@ -0,0 +1,63 @@
+import { ReactNode } from 'react';
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/types.ts:1]: Please review this file

File: packages/big-design/src/hooks/useEventCallback.ts [hunk 17]
```diff
@@ -1,16 +1,16 @@
 import { useCallback, useEffect, useRef } from 'react';
```
Comment (MariaJose) [inline packages/big-design/src/hooks/useEventCallback.ts:1]: Please review this file

**Review by MariaJose — COMMENTED** [review]
Note:  I added a comment to the files that are necessary to review. You can ignore the ones that don't have since they are just a copy paste form the Table component.

File: packages/big-design/src/components/TableNext/HeaderCell/HeaderCell.tsx [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+      return (
+        <Tooltip
+          id={tooltipId}
+          placement="right"
+          trigger={
+            <Box as="span" marginLeft="xxSmall">
+              <BaselineHelpIcon
+                aria-describedby={tooltipId}
+                size="medium"
+                title="Hover or focus for additional context."
```
Comment (bc-juanvasquez) [inline packages/big-design/src/components/TableNext/HeaderCell/HeaderCell.tsx:80]: `title` need `getText`?

Comment (MariaJose) [reply packages/big-design/src/components/TableNext/HeaderCell/HeaderCell.tsx:80]: Good call out! I think we don't translate texts in BD for the moment, however, this is one of our future plans. If I recall correctly the Design team created a doc talking about this. So might be good to wait until we decided what's the next step?

File: packages/big-design/src/components/TableNext/DataCell/styled.tsx [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+`;
+
+export const StyledTableDataCheckbox = styled(StyledTableDataCell)`
+  ${withTableColumnDisplay()}
+
+  background-color: ${({ theme }) => theme.colors.white};
+  padding: ${({ theme }) => `0 ${theme.spacing.small}`};
+
+  &:first-of-type {
+    padding-left: ${({ theme }) => theme.spacing.xLarge};
```
Comment (bc-juanvasquez) [inline packages/big-design/src/components/TableNext/DataCell/styled.tsx:57]: was this style being applied as well?

File: packages/big-design/src/components/TableNext/Row/Row.tsx [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    return null;
+  };
+
+  const renderExpandedIconCell = () => {
+    if (isExpandable && isParentRow && hasChildrenRows) {
+      return (
+        <DataCell
+          cellPadding={isSelectable || isDraggable ? 0 : undefined}
```
Comment (bc-juanvasquez) [inline packages/big-design/src/components/TableNext/Row/Row.tsx:125]: can make this boolean test a `const`

File: packages/big-design/src/components/TableNext/HeaderCell/HeaderCell.tsx (hunk 18 above)
Comment (rtalvarez) [reply packages/big-design/src/components/TableNext/HeaderCell/HeaderCell.tsx:80]: Good callout, but yeah lets not worry too much about gettext for now

**Review by rtalvarez — COMMENTED** [review]
Since this PR is quite big, I'll devote some focused time with @chanceaclark to get it properly reviewed on Wednesday.

File: packages/big-design/src/components/TableNext/Row/useRowState.ts [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+      return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
+    });
+
+  const someChildrenRowsSelected: boolean =
+    isExpandable &&
+    childrenRows.some((_childRow, childRowIndex) => {
+      return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
+    });
+
+  const label = isSelected ? `Selected` : `Unselected`;
```
Comment (bc-juanvasquez) [inline packages/big-design/src/components/TableNext/Row/useRowState.ts:58]: `gettext`?

File: packages/big-design/src/components/TableNext/Row/useRowState.ts [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    if (onExpandedRow) {
+      onExpandedRow(parentRowIndex ?? null);
+    }
+  };
+
+  const hasChildrenRows = childrenRows.length > 0;
+
+  const allChildrenRowsSelected: boolean =
+    isExpandable &&
+    childrenRows.every((_childRow, childRowIndex) => {
```
Comment (bc-juanvasquez) [inline packages/big-design/src/components/TableNext/Row/useRowState.ts:48]: could create a helper function for this and pass in an argument for the 'filter' function. (`some`, `every`)  since the logic is the same for `allChildrenRowsSelected` and `someChildrenRowsSelected`

File: packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts (hunk 13 above)
Comment (bc-juanvasquez) [reply packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts:1]: i think if we change `selectedItems` data structure, this logic could be simplified. I'd like to hear what others have to think about that tho 🤔

File: packages/big-design/src/components/TableNext/Actions/Actions.tsx [hunk 23]
```diff
@@ -14,23 +14,28 @@ export interface ActionsProps<T> {
   forwardedRef: RefObject<HTMLDivElement>;
   itemName?: string;
   items: T[];
+  isExpandable: boolean;
+  isSelectable: boolean;
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/Actions/Actions.tsx:18]: I'll remove this one, I think this isn't necessary

File: packages/big-design/src/components/TableNext/types.ts [hunk 24]
```diff
@@ -44,13 +49,15 @@ export interface TableProps<T> extends React.TableHTMLAttributes<HTMLTableElemen
   actions?: React.ReactNode;
   columns: Array<TableColumn<T>>;
   emptyComponent?: React.ReactElement;
+  expandedRowSelector?: (item: T) => T[] | undefined;
```
Comment (MariaJose) [inline packages/big-design/src/components/TableNext/types.ts:52]: I think should be better to put `expandedRowSelector` inside `TableExpandable` 🤔

File: packages/big-design/src/components/TableNext/DataCell/styled.tsx (hunk 19 above)
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/DataCell/styled.tsx:57]: Just re-check, and this one is used 👍🏼

File: packages/big-design/src/components/TableNext/Row/useRowState.ts (hunk 22 above)
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/Row/useRowState.ts:48]: Tried to create the helper function, but I think is now looking too bulky or distracting? 🤔  was there another way to implement it that I might be missing?

```ts
Could be outside the hook: 
const childrenRowsSelected = <T>(
  mode: 'every' | 'some',
  isExpandable: boolean,
  childrenRows: T[],
  parentRowIndex: number,
  selectedItems: TableSelectable['selectedItems'],
) => {
  if (!isExpandable) {
    return false;
  }

  if (mode === 'every') {
    return childrenRows.every((_childRow, childRowIndex) => {
      return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
    });
  }

  return childrenRows.some((_childRow, childRowIndex) => {
    return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
  });
};

or Inside the hook:
  const childrenRowsSelected = <T>(
    mode: 'every' | 'some',
  ) => {
    if (!isExpandable) {
      return false;
    }
  
    if (mode === 'every') {
      return childrenRows.every((_childRow, childRowIndex) => {
        return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
      });
    }
  
    return childrenRows.some((_childRow, childRowIndex) => {
      return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
    });
  };
```

Comment (bc-juanvasquez) [reply packages/big-design/src/components/TableNext/Row/useRowState.ts:48]: `const filterChildrenRows = (type) => {
	const dictionary = {
  	'some': Array.some,
    'every': Array.every,
    //extensible to other filters
  }
  const filterFunction = dictionary[type] ?? null;
  const cb = (el, childRowIndex, childrenRows) => selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined
  
  return filterFunction ? filterFunction(cb) : false;
}

const allChildrenRowsSelected = isExpandable && filterChildrenRows('every')`

 was thinking something along these lines, but not sure if it will work 🤔

File: packages/big-design/src/components/TableNext/Row/styled.tsx [hunk 25]
```diff
@@ -20,4 +20,12 @@ export const StyledTableRow = styled.tr<StyledTableRowProps>`
   }
 `;
 
+export const StyledExpandedIcon = styled.button`
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/Row/styled.tsx:23]: Is there a reason why we are not using our private [`StylableButton`](https://github.com/bigcommerce/big-design/blob/master/packages/big-design/src/components/Button/Button.tsx#L50)? That way we can use our `Button` props.

File: packages/big-design/src/components/TableNext/Row/useRowState.ts (hunk 22 above)
Comment (chanceaclark) [reply packages/big-design/src/components/TableNext/Row/useRowState.ts:48]: I'd avoid using a dictionary to point to a prototype method. I'm actually more fond of leaving it as is since it's a bit more readable. If we did want to make it a little better, we could rip it outside of the hook and have it be a bit more generic. But I would only do that if we are going to reuse it in other places 🤷 
```ts
// "All" example
const areAllRowsSelected = (rows, selectedItems) => {
  return rows.some((row, index) => {
      const parentIndex = getParentRowIndex(row); // TODO: Define utility function;

      return selectedItems[`${parentIndex}.${index}`] !== undefined;
  });
}
```

File: packages/big-design/src/components/TableNext/DataCell/DataCell.tsx [hunk 26]
```diff
@@ -9,11 +9,13 @@ export interface DataCellProps
...
   align?: 'left' | 'center' | 'right';
   children?: React.ReactNode;
+  isExpandable?: boolean;
   isCheckbox?: boolean;
   verticalAlign?: 'top' | 'middle';
   width?: number | string;
   withBorder?: boolean;
-  withPadding?: boolean;
+  withVerticalPadding?: boolean;
+  withHorizontalPadding?: boolean;
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/DataCell/DataCell.tsx:18]: Instead of defining new props, why don't we try and extend off of our `PaddingProps`? May not be possible but worth a shot.

File: packages/big-design/src/components/TableNext/Row/Row.tsx [hunk 27]
```diff
@@ -1,60 +1,165 @@
...
-
-interface PrivateProps {
-  forwardedRef?: React.Ref<HTMLTableRowElement>;
+  onExpandedRow?(parentRowIndex: number | null): void;
+  onItemSelect?(
+    childRowIndex: number | null,
+    index: number,
+    childrenRows: T[],
+    isParentRow: boolean,
+    isExpandable: boolean,
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/Row/Row.tsx:39]: Is there a way where we can reduce the amount of arguments passed? Anything greater than 2-3 arguments can be quite cumbersome. Could this be a single interface a pass in instead (like it was before)?

File: packages/big-design/src/components/TableNext/Row/Row.tsx [hunk 28]
```diff
@@ -1,60 +1,165 @@
...
+
+  const renderExtraCellsForParentRow = () => {
+    const extraDataCells: ReactNode[] = [];
+
+    if (!hasChildrenRows) {
+      extraDataCells.push(<DataCell key={`parent-extra-cell-${parentRowIndex}-1`} />);
+    }
+
+    return extraDataCells;
+  };
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/Row/Row.tsx:140]: We could simplify this a bit instead of relying on array mutations
```suggestion
  const renderExtraCellsForParentRow = (): ReactNode[] => {
    if (!hasChildrenRows) {
      return [<DataCell key={`parent-extra-cell-${parentRowIndex}-1`} />];
    }

    return [];
  };
```

File: packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+    return;
+  }
+
+  const isSelectedChild = selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
+
+  if (isSelectedChild) {
+    delete selectedItems[`${parentRowIndex}.${childRowIndex}`];
+
+    if (isTheOnlySelectedChildRow) {
+      delete selectedItems[`${parentRowIndex}`];
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts:61]: 😬 Can we try and find a better way for this instead of mutating the array? Mutations are side-effects and are hard to track down if there are any issues with it. It's better to return an immutable array instead.

File: packages/big-design/src/components/TableNext/hooks/useExpandable/useExpandable.ts [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+      return;
+    }
+
+    const { onExpandedChange } = expandable;
+
+    const isExpandedRow = expandedRows[parentRowIndex] !== undefined;
+    const newExpandedRows = { ...expandedRows };
+
+    if (isExpandedRow) {
+      delete newExpandedRows[parentRowIndex];
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/hooks/useExpandable/useExpandable.ts:21]: Same here.

File: packages/big-design/src/components/TableNext/SelectAll/helpers.ts [hunk 31]
```diff
@@ -0,0 +1,115 @@
...
+  return expandedRowSelector(parentRow) ?? [];
+}
+
+export function areAllInPageSelected<T>(
+  isExpandable: boolean,
+  items: T[],
+  selectedItems: TableSelectable['selectedItems'],
+  expandedRowSelector?: TableExpandable<T>['expandedRowSelector'],
+) {
+  return (
```
Comment (rtalvarez) [inline packages/big-design/src/components/TableNext/SelectAll/helpers.ts:35]: Hmm I think these `return`s in this file could be tweaked a bit. ATM I think it's hard to follow them because theres a `return` -> `items.every` -> more `return` branches -> nested `childrenRows.every` -> `return` branches

Maybe pulling this into logical helper functions? Or moving some stuff around 🤔

File: packages/big-design/src/components/TableNext/SelectAll/useSelectAllState.ts [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+
+    if (allInPageSelected) {
+      selectAll(isExpandable, items, newSelectedItems, expandedRowSelector);
+
+      return onChange(newSelectedItems);
+    }
+
+    unselectAll(isExpandable, items, newSelectedItems, expandedRowSelector);
+
+    return onChange(newSelectedItems);
```
Comment (jorgemoya) [inline packages/big-design/src/components/TableNext/SelectAll/useSelectAllState.ts:58]: Why does it look like the `selectAll` and `unselectAll` are reversed 😆 . I'm sure it's working fine, just having hard time in understanding why we wouldn't `unselectAll` if `allInPageSelected`?

File: packages/big-design/src/components/TableNext/mixins/display/display.tsx [hunk 33]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+const getSimpleDisplay = (
+  displayProp: string | number,
+  cssKey: string,
+): FlattenSimpleInterpolation => css`
+  ${cssKey}: ${displayProp}
+`;
+
+const getResponsiveDisplay: TableColumnDisplayOverload = (
```
Comment (jorgemoya) [inline packages/big-design/src/components/TableNext/mixins/display/display.tsx:33]: Wondering why we are not using the mixins that already exist in `src/mixins/display`?

**Review by jorgemoya — COMMENTED** [review]
Looking good!

File: packages/big-design/src/components/TableNext/Actions/styled.tsx [hunk 34]
```diff
@@ -0,0 +1,25 @@
+import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
+import styled, { css } from 'styled-components';
+
+import { FlexProps } from '../../Flex';
+import { withFlexedContainer } from '../../Flex/withFlex';
+
+export const StyledFlex = styled.div<FlexProps & { stickyHeader?: boolean }>`
+  ${withFlexedContainer()}
```
[the remark this answers is not in the document]
Comment (chanceaclark) [reply packages/big-design/src/components/TableNext/Actions/styled.tsx:8]: I guess it really depends on the situation. If we extend via `styled(Flex)` we introduce a lot more props than just the `FlexProps`; this would introduce `BoxProps`, `PaddingProps`, `MarginProps`, & `DisplayProps`. If we do it the `styled.div<FlexProps>` way we can narrow the props to what we want without having to filter out the unwanted props.

File: packages/big-design/src/components/TableNext/Row/styled.tsx (hunk 25 above)
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/Row/styled.tsx:23]: Tried with the public one, but didn't know we could use that one ^  😓 , so decided to create a new one because I was having issues with the styles, but good to know that this one is more flexible.

Also the `StylableButton` adds styles when button is active, focused,  and hover. I guess I'm gonna leave these styles for now to have better accessibility even-though that is not specified in the designs, I think we could change it later in case this is not required.

Screen Shot 2022-07-21 at 1 33 19 PM

File: packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts (hunk 29 above)
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts:61]: I think I'm not mutating the object. Before calling this helper fn I'm creating a copy and passed the copy to the helper function. Reference: packages/big-design/src/components/TableNext/hooks/useSelectable/useSelectable.ts

I think it might be confused because I naming this prop as selectedItems, probably I should name it differently, or make the copy inside the this helper function to prevent confusion.

File: packages/big-design/src/components/TableNext/DataCell/DataCell.tsx (hunk 26 above)
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/DataCell/DataCell.tsx:18]: Tried to extended it, but things got messed up with styles, I'll leave this suggestion until the end.

File: packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts (hunk 29 above)
Comment (chanceaclark) [reply packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts:61]: You are still mutating a passed object and mutating it's contents. Could we try and make this function a bit smaller, reduce the amount of arguments and have it do one thing: take an object and return a new one. Smaller functions are highly preferred.

```ts
const selectChildRow = (selectedItems: TableSelectable['selectedItems'], childRowIndex: number, ): TableSelectable['selectedItems'] => {
  const parentRowIndex = getParentIndex(childRowIndex);
  const isSelectedParent = selectedItems[parentRowIndex] !== undefined;

  if (!isSelectedParent) {
    selectedItems[`${parentRowIndex}`] = true;
    selectedItems[`${parentRowIndex}.${childRowIndex}`] = true;

    return {
      ...selectedItems,
      [`${parentRowIndex}`]: true,
      [`${parentRowIndex}.${childRowIndex}`] = true;
    };
  }

  // ...
}

// Usage:

const newSelectedItems = selectChildRow(selectedItems, childRowIndex);

onSelectionChange(newSelectedItems);
```

Instead of deleting keys could we just update the value to false?

Comment (deini) [reply packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts:61]: This example is still mutating, left-over copy paste?

File: packages/big-design/src/components/TableNext/SelectAll/useSelectAllState.ts (hunk 32 above)
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/SelectAll/useSelectAllState.ts:58]: lol, good catch, thanks! 😓  I swapped the names

File: packages/big-design/src/components/TableNext/mixins/display/types.ts [hunk 35]
```diff
@@ -0,0 +1,16 @@
+import { ThemeInterface } from '@bigcommerce/big-design-theme';
+import { FlattenSimpleInterpolation } from 'styled-components';
+
+import { ResponsiveProp } from '../../../../types';
+
+type TableColumnDisplayProp = ResponsiveProp<'table-cell' | 'none'>;
+
+export type TableColumnDisplayProps = Partial<{
+  display: TableColumnDisplayProp;
+}>;
```
[the remark this answers is not in the document]
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/mixins/display/types.ts:10]: I agree, we could use `?` instead of `Partial`. I just gonna change `type` to `interface` since it's causing eslint errors `Use an `interface` instead of a `type`

would be like this:
```
export interface TableColumnDisplayProps {
  display?: TableColumnDisplayProp;
}
```

File: packages/big-design/src/components/TableNext/mixins/display/display.tsx (hunk 33 above)
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/mixins/display/display.tsx:33]: Not sure tbh, this was a copy paste from the Table component, it was like that before. But I'll take a look and see what I can do

File: packages/big-design/src/components/TableNext/SelectAll/helpers.ts (hunk 31 above)
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/SelectAll/helpers.ts:35]: That's true, I created a helper function so it can be easier to follow 👍🏼

File: packages/big-design/src/components/TableNext/mixins/display/display.tsx (hunk 33 above)
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/mixins/display/display.tsx:33]: I already checked and I think is like that because the types are different, so I imagine that even-though it's almost the same, it was better to keep it separately to make it more readable ? 🤔  I think I'm gonna leave it like that and we can create another ticket to handle that scope.

**Comment by MariaJose** [comment]
♻️  @bigcommerce/channels-fe

**Review by jorgemoya — APPROVED** [review]
lgtm!

File: packages/big-design/src/components/TableNext/SelectAll/useSelectAllState.ts [hunk 36]
```diff
@@ -0,0 +1,69 @@
+import { TableExpandable, TableSelectable } from '../types';
+
+import {
+  areAllInPageSelected,
+  areSomeInPageSelected,
+  getTotalSelectedItems,
+  selectAll,
+  // unselectAll,
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/SelectAll/useSelectAllState.ts:8]: Unused?

File: packages/big-design/src/components/TableNext/hooks/useSelectable/useSelectable.ts [hunk 37]
```diff
@@ -0,0 +1,83 @@
...
+    },
+  );
+
+  useEffect(() => {
+    if (selectable) {
+      const newSelectedItems = { ...selectable.selectedItems };
+
+      setSelectedItems(newSelectedItems);
+    }
+  }, [selectable, selectableConditionalDep]);
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/hooks/useSelectable/useSelectable.ts:76]: `const selectableConditionalDep = selectable ? selectable.selectedItems : null;` is generally not ideal. We should be able to use optional chaining instead.
```suggestion
  useEffect(() => {
    if (selectable?.selectedItems) {
      setSelectedItems({ ...selectable.selectedItems });
    }
  }, [selectable?.selectedItems]);
```

File: packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts [hunk 38]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+export function getTotalSelectedChildRows<T>({
+  childrenRows,
+  parentRowIndex,
+  selectedItems,
+}: SelectRowArg<T>) {
+  return (
+    childrenRows !== undefined &&
+    childrenRows.reduce((acc, _childRow, childRowIndex) => {
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts:201]: Can we use optional chaining instead `childrenRows?.reduce`?

File: packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts [hunk 39]
```diff
@@ -0,0 +1,209 @@
...
+}: SelectRowArg<T>) {
+  const newSelectedItems = { ...selectedItems };
+
+  delete newSelectedItems[`${parentRowIndex}.${childRowIndex}`];
+
+  if (isTheOnlySelectedChildRow) {
+    delete newSelectedItems[`${parentRowIndex}`];
+  }
+
+  return newSelectedItems;
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts:191]: Again, we probably shouldn't mutate even if we are creating a new object/array. Maybe we can destructure where we delete instead?
```ts
const parentChildIndex = `${parentRowIndex}.${childRowIndex}`;
const { [parentChildIndex]: _child, ...removedChildSelectedItems } = selectedItems;

if (isTheOnlySelectedChildRow) {
  const parentIndex = `${parentRowIndex}`;
  const { [parentIndex]: _parent, ...removedParentSelectedItems } = removedChildSelectedItems;
   
  return removedParentSelectedItems;
}

return removedChildSelectedItems;
```

**Comment by MariaJose** [comment]
♻️

**Comment by MariaJose** [comment]
♻️  @bigcommerce/channels-fe

File: packages/big-design/src/components/TableNext/SelectAll/helpers.ts [hunk 40]
```diff
@@ -0,0 +1,127 @@
...
+      childrenRows.forEach((_childRow, childRowIndex) => {
+        if (newSelectedItems[`${parentRowIndex}.${childRowIndex}`] === undefined) {
+          newSelectedItems[`${parentRowIndex}.${childRowIndex}`] = true;
+        }
+      });
+    }
+  });
+
+  return newSelectedItems;
+}
```
Comment (chanceaclark) [inline packages/big-design/src/components/TableNext/SelectAll/helpers.ts:127]: Do we need to know about previous selected items or can we just build a new object from scratch? This may not be 100% right, but this was what I was thinking:
```suggestion
export function selectAll<T>({
  expandedRowSelector,
  isExpandable,
  items,
}: SelectAllRowsArg<T>): TableSelectable['selectedItems'] {
  return Object.fromEntries(items.map((parent, parentIndex) => {
    const children = getChildrenRows(parent, expandedRowSelector);

    const selectedChildren = isExpandable ? children.map((_child, childIndex) => {
        return [`${parentIndex}.${childIndex}`, true];
    }) : [];

    return [
        [`${parentIndex}`, true],
        ...selectedChildren,
    ];
  }));
}
```

**Review by chanceaclark — COMMENTED** [review]
Few more comments, otherwise looking really good.

File: packages/big-design/src/components/TableNext/RowContainer/RowContainer.tsx [hunk 41]
```diff
@@ -0,0 +1,100 @@
...
+  expandedRowSelector,
+  getItemKey,
+  onItemSelect,
+  onExpandedRow,
+  selectedItems,
+  ...rest
+}: InternalRowContainerProps<T> & PrivateProps) => {
+  const isParentRowSelected = selectedItems[parentRowIndex] !== undefined;
+  const isExpanded = expandedRows[parentRowIndex] !== undefined;
+  const childrenRows: T[] | undefined = expandedRowSelector ? expandedRowSelector(item) : [];
```
[the remark this answers is not in the document]
Comment (MariaJose) [reply packages/big-design/src/components/TableNext/RowContainer/RowContainer.tsx:39]: That's true, looks better 👍🏼

**Comment by MariaJose** [comment]
♻️

**Review by jorgemoya — COMMENTED** [review]
lgtm!
